[BEAM-1974] Add Metrics user-oriented documentation (#9328)
apache/beam: website/src/documentation/programming-guide.md

```py
{% github_sample /apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets_test.py tag:model_other_composite_triggers
%}```

## 9. Metrics {#metrics}

> The Beam model provides a way to use different types of metrics in a user pipeline.  There could be 

Reviewer: ```suggestion
In the Beam model, metrics provide some insight into the current state of a user pipeline, potentially while the pipeline is running.  There could be 
```
(Just a suggestion -- the first sentence should describe why metrics are interesting.  The examples are a good illustration!)

*   ...and so on.

### 9.1 The main concepts of Beam metrics
*   **Named**. Each metric has a name which consists of a namespace and an actual name. The 
    namespace can be used to differentiate between multiple metrics with the same name and also 
    allows querying for all metrics within a specific namespace. 
*   **Scoped**. Each metric is reported against a specific step in the pipeline, indicating what 
    code was running when the metric was incremented.
*   **Dynamically Created**. Metrics may be created during runtime without pre-declaring them, in 
    much the same way a logger could be created. This makes it easier to produce metrics in utility 
    code and have them usefully reported. 
*   **Degrade Gracefully**. If a runner doesn’t support some part of reporting metrics, the 
    fallback behavior is to drop the metric updates rather than failing the pipeline. If a runner 
    doesn’t support some part of querying metrics, the runner will not return the associated data.

Reported metrics are implicitly scoped to the transform within the pipeline that reported them. 
This allows reporting the same metric name in multiple places and identifying the value each 

> transform reported, as well as aggregating the metric across

Reviewer: ```suggestion
transform reported, as well as aggregating the metric across the entire pipeline.
```

```java
Gauge gauge = Metrics.gauge( "namespace", "gauge1");

@ProcessElement
public void processElement(ProcessContext context) {
  Integer element = context.element();
  // create a gauge (latest value received) of the values 
  gauge.set(element);
  ...
}
```

### 9.3 Querying metrics {#querying-metrics}
`PipelineResult` has a method `metrics()` which returns a `MetricResults` object that allows 
accessing metrics. The main method available in `MetricResults` allows querying for all metrics 
matching a given filter.

```java
public interface PipelineResult {
  MetricResults metrics();
}

public abstract class MetricResults {
  public abstract MetricQueryResults queryMetrics(@Nullable MetricsFilter filter);
}

public interface MetricQueryResults {
  Iterable<MetricResult<Long>> getCounters();
  Iterable<MetricResult<DistributionResult>> getDistributions();
  Iterable<MetricResult<GaugeResult>> getGauges();
}

public interface MetricResult<T> {
  MetricName getName();
  String getStep();
  T getCommitted();
  T getAttempted();
}
```

### 9.4 Using metrics in pipeline {#using-metrics}

> Below, there is a simple example of how to use `Counter` metric in user pipeline.

Reviewer: ```suggestion
Below, there is a simple example of how to use a `Counter` metric in a user pipeline.
```

```java
// creating a pipeline with custom metrics DoFn
pipeline
    .apply(...)
    .apply(ParDo.of(new MyMetricsDoFn()));

pipelineResult = pipeline.run().waitUntilFinish(...);

// query metric by namespace and metric name
MetricQueryResults metrics =
    pipelineResult
        .metrics()
        .queryMetrics(
            MetricsFilter.builder()
                .addNameFilter(MetricNameFilter.named("namespace", "counter1"))

>                 .build());

Reviewer: ```suggestion
                .build());
// Find and print the queried counter:
for (MetricResult<Long> r: metrics.getCounters()) {
  System.out.println(r.getName() + ":" + r.getAttempted());
}
```

matching a given filter.

```java
public interface PipelineResult {
  MetricResults metrics();
}

public abstract class MetricResults {
  public abstract MetricQueryResults queryMetrics(@Nullable MetricsFilter filter);
}

public interface MetricQueryResults {
  Iterable<MetricResult<Long>> getCounters();
  Iterable<MetricResult<DistributionResult>> getDistributions();
  Iterable<MetricResult<GaugeResult>> getGauges();
}

public interface MetricResult<T> {
  MetricName getName();
  String getStep();
  T getCommitted();
  T getAttempted();
}
```

### 9.4 Using metrics in pipeline {#using-metrics}
Below, there is a simple example of how to use a `Counter` metric in a user pipeline.

```java
// creating a pipeline with custom metrics DoFn
pipeline
    .apply(...)
    .apply(ParDo.of(new MyMetricsDoFn()));

pipelineResult = pipeline.run().waitUntilFinish(...);

// query metric by namespace and metric name
MetricQueryResults metrics =
    pipelineResult
        .metrics()
        .queryMetrics(
            MetricsFilter.builder()
                .addNameFilter(MetricNameFilter.named("namespace", "counter1"))
                .build());

// Find and print the queried counter:

> for (MetricResult<Long> counter: metrics.getCounters()) {

Author: I added a comment since I think it will worth to show that we can iterate over all counters.

> for (MetricResult<Long> counter: metrics.getCounters()) {

Reviewer 2: Yes, but what troubles me is actually the "find" word, because we are not searching anything. I think the comments are not saying what they do. I would put:

- "Request the metric called "counter" in namespace called "namespace" instead of "query metric by namespace and metric "
- "Print the metric value (there should be only one line because there is only one metric called "counter1" in the namespace called "namespace")" instead of "Find and print the queried counter, it's expected that only one line will be name:printed"

> for (MetricResult<Long> counter: metrics.getCounters()) {

Author: Ok, I agree, thanks

  // create a gauge (latest value received) of the values 
  gauge.set(element);
  ...
}
```

### 9.3 Querying metrics {#querying-metrics}
`PipelineResult` has a method `metrics()` which returns a `MetricResults` object that allows 
accessing metrics. The main method available in `MetricResults` allows querying for all metrics 
matching a given filter.

```java
public interface PipelineResult {
  MetricResults metrics();
}

public abstract class MetricResults {
  public abstract MetricQueryResults queryMetrics(@Nullable MetricsFilter filter);
}

public interface MetricQueryResults {
  Iterable<MetricResult<Long>> getCounters();
  Iterable<MetricResult<DistributionResult>> getDistributions();
  Iterable<MetricResult<GaugeResult>> getGauges();
}

public interface MetricResult<T> {
  MetricName getName();
  String getStep();
  T getCommitted();
  T getAttempted();
}
```

### 9.4 Using metrics in pipeline {#using-metrics}
Below, there is a simple example of how to use a `Counter` metric in a user pipeline.

```java
// creating a pipeline with custom metrics DoFn
pipeline
    .apply(...)
    .apply(ParDo.of(new MyMetricsDoFn()));

pipelineResult = pipeline.run().waitUntilFinish(...);

> // request the metric, called "counter1", in namespace, called "namespace"

Reviewer 2: remove the 3 ","

MetricQueryResults metrics =
    pipelineResult
        .metrics()
        .queryMetrics(
            MetricsFilter.builder()
                .addNameFilter(MetricNameFilter.named("namespace", "counter1"))
                .build());

> // print the metric value - there should be only one line because there is only one metric, 

Reviewer 2: remove ","

> // called "counter1", in the namespace, called "namespace"

Reviewer 2: remove the 2 ","
